crates/iceberg/src/arrow.rs

```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+                            ErrorKind::DataInvalid,
+                            format!(
+                                "Leave column {:?} in schema doesn't have field id",
+                                field_type
+                            ),
+                        ));
+                    }
+                    column_map.insert(basic_info.id(), idx);
+                }
+                ParquetType::GroupType { .. } => {
+                    return Err(Error::new(
+                        ErrorKind::DataInvalid,
+                        format!(
+                            "Leave column in schema should be primitive type but got {:?}",
+                            field_type
+                        ),
+                    ));
+                }
+            };
+        }
+
+        Ok(column_map)
+    }
+}
+
+/// A visitor to collect field ids from bound predicates.
+struct CollectFieldIdVisitor {
+    field_ids: Vec<i32>,
+}
+
+impl BoundPredicateVisitor for CollectFieldIdVisitor {
+    type T = ();
+    type U = ();
+
+    fn and(&mut self, _predicates: Vec<Self::T>) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn or(&mut self, _predicates: Vec<Self::T>) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn not(&mut self, _predicate: Self::T) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn visit_always_true(&mut self) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn visit_always_false(&mut self) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn visit_unary(&mut self, predicate: &UnaryExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn visit_binary(&mut self, predicate: &BinaryExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn visit_set(&mut self, predicate: &SetExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn bound_reference(&mut self, reference: &BoundReference) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+}
+
+struct PredicateConverter<'a> {
+    pub columns: &'a Vec<usize>,
+    pub projection_mask: ProjectionMask,
+    pub parquet_schema: &'a SchemaDescriptor,
+    pub column_map: &'a HashMap<i32, usize>,
+}
+
+fn get_arrow_datum(datum: &Datum) -> Box<dyn ArrowDatum> {
+    match datum.literal() {
+        PrimitiveLiteral::Boolean(value) => Box::new(BooleanArray::new_scalar(*value)),
+        PrimitiveLiteral::Int(value) => Box::new(Int32Array::new_scalar(*value)),
+        PrimitiveLiteral::Long(value) => Box::new(Int64Array::new_scalar(*value)),
+        PrimitiveLiteral::Float(value) => Box::new(Float32Array::new_scalar(value.as_f32())),
+        PrimitiveLiteral::Double(value) => Box::new(Float64Array::new_scalar(value.as_f64())),
+        _ => todo!("Unsupported literal type"),
```

Comment:
We should not panic here, but throws error.

Reply:
Currently I cannot return an error from `get_arrow_datum` to propagate an error. The closure needs to get the ownership of all referenced variables. But we cannot call `get_arrow_datum` outside the closure to propagate the error and pass `Datum` into the closure because `Datum` is not implemented with `Send`. I proposed https://github.com/apache/arrow-rs/pull/5587 and see if we can get it.